flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala

```diff
@@ -236,6 +236,32 @@ case class Aggregate(
   }
 }
 
+case class SetMinus(left: LogicalNode, right: LogicalNode, all: Boolean) extends BinaryNode {
+  override def output: Seq[Attribute] = left.output
+
+  override protected[logical] def construct(relBuilder: RelBuilder): RelBuilder = {
+    left.construct(relBuilder)
+    right.construct(relBuilder)
+    relBuilder.minus(all)
+  }
+
+  override def validate(tableEnv: TableEnvironment): LogicalNode = {
+    val resolvedMinus = super.validate(tableEnv).asInstanceOf[SetMinus]
+    if (left.output.length != right.output.length) {
+      failValidation(s"Set minus two table of different column sizes:" +
+        s" ${left.output.size} and ${right.output.size}")
+    }
+    val sameSchema = left.output.zip(right.output).forall { case (l, r) =>
+      l.resultType == r.resultType && l.name == r.name }
```

Comment:
I think @wuchong refers exactly to the last case. If number of fields and field types are identical, it should be possible to do a `EXCEPT` even if the field names are not the same.
The situation for `UNION` is a bit special due to Flink's internals and the way the `RowTypeInfo` is implemented but I think we can remove that restriction in the future. 

So, we must keep the checks for number of fields and field types but can remove the check of the field names, IMO.